qemu/tests/hotplug_mem_share_discard_data.py

```diff
@@ -0,0 +1,93 @@
+from avocado.core import exceptions
+from avocado.utils import process
+
+from virttest import error_context
+from virttest import utils_misc
+
+from virttest.utils_numeric import normalize_data_size
+from virttest.utils_test.qemu import MemoryHotplugTest
+from virttest.staging import utils_memory
+
+
+@error_context.context_aware
+def run(test, params, env):
+    """
+    Memory share and discard-data hotplug test
+    1) Setup hugepages
+    2) Mount /mnt/kvm_hugepage
+    3) Boot guest with one numa node
+    4) Hotplug 1G memory
+    5) Check memory
+    6) Unplug memory
+    7) Check memory
+    8) Check results for discard-data value
+
+    :param test: QEMU test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment
+    """
+
+    timeout = float(params.get("login_timeout", 240))
+
+    vm = env.get_vm(params["main_vm"])
+    vm.verify_alive()
+    session = vm.wait_for_login(timeout=timeout)
+
+    mem_name = params["target_mems"]
+    hp_size = utils_memory.read_from_meminfo("Hugepagesize")
+    hp_free = int(utils_memory.read_from_meminfo('HugePages_Free'))
+    size_target_mem = params["size_mem_%s" % mem_name]
+
+    hp_total = int(utils_memory.read_from_meminfo("HugePages_Total"))
+    hp_plugged = int(normalize_data_size(size_target_mem, "K"))
+    hp_target = int(hp_plugged / hp_size)\
+        + int(hp_total)
+    process.system("echo %s > /proc/sys/vm/nr_hugepages" % hp_target, shell=True)
+
+    is_discard_data = params.get("discard-data_plug1", "yes")
+    is_backend_file = params["backend_mem_plug1"]
+
+    hotplug_test = MemoryHotplugTest(test, params, env)
+    hotplug_test.hotplug_memory(vm, mem_name)
+    hotplug_test.check_memory(vm)
+
+    # Checks if hotplugged device has share: True option
+    if utils_misc.compare_qemu_version(6, 1, 0, is_rhev=False) is True:
+        mem_dev_name = "mem-%s" % params["target_mems"]
+        query_list = vm.monitor.info("memdev")
+        for dev in query_list:
+            error_context.context("device id: %s has share: %s" % (dev['id'], dev['share']), test.log.debug)
+            if dev['id'] == mem_dev_name and dev['share'] is not True:
+                raise exceptions.TestFail("dev %s has share option: %s" % (dev['id'], dev['share']))
+
+    hotplug_test.unplug_memory(vm, mem_name)
+    hotplug_test.check_memory(vm)
+
+    session.close()
+    error_context.context("Shutdown guest...")
+    vm.destroy()
```

Comment:
For manual case, it compare HugePages_Total and HugePages_Free before destroy vm, but automation do it after destroy vm.  Are they always the same ?   
The hotplug mem with discard-data ‘on’ can not be released even if vm be destroyed, until the `test_file` be deleted ?

Reply:
Well in this case I think you mean with 'discard-data = off' because data (the file) **will not be discarded**.

And yes, you need to manually delete the file, because qemu won't do it for discard-data=off. So for testing if this feature is working, there is no need of having the vm working, that's why it is destroyed. In the discard-data=on case total memory will be equals to the free one so it can be validated with the vm destroyed too.